src/Microsoft.VisualStudio.Threading/JoinableTask.cs

```diff
@@ -200,28 +200,8 @@ public bool IsCompleted
         {
             get
             {
-                using (this.JoinableTaskContext.NoMessagePumpSynchronizationContext.Apply())
-                {
-                    lock (this.JoinableTaskContext.SyncContextLock)
-                    {
-                        if (!this.IsCompleteRequested)
-                        {
-                            return false;
-                        }
-
-                        if (this.mainThreadQueue is object && !this.mainThreadQueue.IsCompleted)
-                        {
-                            return false;
-                        }
-
-                        if (this.threadPoolQueue is object && !this.threadPoolQueue.IsCompleted)
-                        {
-                            return false;
-                        }
-
-                        return true;
-                    }
-                }
+                return this.TryGetTask(out Task? task)
+                    && task.IsCompleted;
```

Comment:
> Why not do the opposite: introduce a new internal member with the new behavior if you need that behavior?

That's exactly what I did for the first commit of this PR. Unfortunately that leaves the type in a state where awaiting the completed `JoinableTask` is efficient but checking `IsCompleted` is not.

> This looks like it produces observably different behavior from a public API.

The change in behavior certainly impacts some of the internal implementation details, but I couldn't find a reason why it would impact external consumers. Based on this, internal usage was moved to an internal member.